sklearn/linear_model/ridge.py

```diff
@@ -995,6 +997,129 @@ def _diag_dot(self, D, B):
+
+        If ``center`` is ``True``, compute
+        (X - X.mean(axis=0)).dot((X - X.mean(axis=0)).T)
+        else X.dot(X.T)
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The input uncentered data.
+        center : bool, default is True
+            Whether or not to remove the mean from ``X``.
+
+        Returns
+        -------
+        gram : ndarray, shape (n_samples, n_samples)
+            The Gram matrix.
+        X_m : ndarray, shape (n_feature,)
+            The mean of ``X`` for each feature.
+        """
+        if not center:
+            # in this case centering has been done in preprocessing
+            # or we are not fitting an intercept.
+            X_mean = np.zeros(X.shape[1], dtype=X.dtype)
+            return safe_sparse_dot(X, X.T, dense_output=True), X_mean
+        # otherwise X is always sparse
+        n_samples = X.shape[0]
+        X_weighted = self._sqrt_sw_matrix.dot(X)
+        X_mean, _ = mean_variance_axis(X_weighted, axis=0)
+        X_mean *= n_samples / self._weight_sum
+        X_mX = self._sqrt_sw[:, None] * safe_sparse_dot(
+            X_mean, X.T, dense_output=True)
+        X_mX_m = np.empty((n_samples, n_samples), dtype=X.dtype)
+        X_mX_m[:, :] = np.dot(X_mean, X_mean)
+        X_mX_m *= self._sqrt_sw
+        X_mX_m *= self._sqrt_sw[:, None]
+        return (safe_sparse_dot(X, X.T, dense_output=True) + X_mX_m
+                - X_mX - X_mX.T, X_mean)
+
+    def _compute_covariance(self, X, center=True):
+        """Computes centered covariance matrix.
+
+        if center is True, compute
+        (X - X.mean(axis=0)).T.dot(X - X.mean(axis=0))
+        else
+        X.T.dot(X)
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The input uncentered data.
+        center : bool, default is True
+            Whether or not to remove the mean from ``X``.
+
+        Returns
+        -------
+        covariance : ndarray, shape (n_features, n_features)
+            The covariance matrix.
+        X_m : ndarray, shape (n_feature,)
+            The mean of ``X`` for each feature.
+        """
+        if not center:
+            # in this case centering has been done in preprocessing
+            # or we are not fitting an intercept.
+            X_mean = np.zeros(X.shape[1], dtype=X.dtype)
+            return safe_sparse_dot(X.T, X, dense_output=True), X_mean
+        # otherwise X is always sparse
+        n_samples = X.shape[0]
+        X_weighted = self._sqrt_sw_matrix.dot(X)
+        X_mean, _ = mean_variance_axis(X_weighted, axis=0)
+        X_mean = X_mean * n_samples / self._weight_sum
+        return (safe_sparse_dot(X.T, X, dense_output=True) -
+                self._weight_sum * np.outer(X_mean, X_mean),
+                X_mean)
+
+    def _sparse_multidot_diag(self, X, A, Xm):
+        """
+        compute the diagonal of (X - Xm).dot(A).dot((X - Xm).T)
```

Comment:
"""Compute the diagonal of (X - X_mean).dot(A).dot((X - X_mean).T)
without explicitly centering `X` when `X` is sparse.
"""